### PR TITLE
[libc++][NFC] Use __construct_at and __destroy_at instead of using preprocessor conditionals

### PR DESCRIPTION
**Pull request by philnik777**

### CODE REVIEW

**Review by ldionne — APPROVED** [review]
LGTM % CI.

File: libcxx/test/std/containers/sequences/vector/vector.cons/copy.move_only.verify.cpp [hunk 1]
```diff
@@ -16,5 +16,5 @@
 
 void f() {
     std::vector<MoveOnly> v;
-    std::vector<MoveOnly> copy = v; // expected-error-re@* {{{{(no matching function for call to 'construct_at')|(call to implicitly-deleted copy constructor of 'MoveOnly')|(call to deleted constructor of 'MoveOnly')}}}}
+    std::vector<MoveOnly> copy = v; // expected-error-re@* {{{{(no matching function for call to '__construct_at')|(call to implicitly-deleted copy constructor of 'MoveOnly')}}}}
```
Comment (EricWF) [inline libcxx/test/std/containers/sequences/vector/vector.cons/copy.move_only.verify.cpp:19]: It's unfortunate that `__construct_at` appears here, but I think that suggests we're testing too much of the compiler message. 

A test should only fail when the behavior under test no longer holds. Is the name part of the behavior under test?

File: libcxx/include/__memory/allocator_traits.h [hunk 2]
```diff
@@ -319,11 +315,7 @@ struct _LIBCPP_TEMPLATE_VIS allocator_traits
         __enable_if_t<!__has_destroy<allocator_type, _Tp*>::value> >
     _LIBCPP_INLINE_VISIBILITY _LIBCPP_CONSTEXPR_SINCE_CXX20
     static void destroy(allocator_type&, _Tp* __p) {
-#if _LIBCPP_STD_VER >= 20
```
Comment (EricWF) [inline libcxx/include/__memory/allocator_traits.h:322]: Why is using `destroy_at` better than simply invoking the destructor? One costs a lot less to instantiate. 
It's  not a customization point, right?

File: libcxx/include/__memory/allocator_traits.h [hunk 3]
```diff
@@ -300,11 +300,7 @@ struct _LIBCPP_TEMPLATE_VIS allocator_traits
         __enable_if_t<!__has_construct<allocator_type, _Tp*, _Args...>::value> >
     _LIBCPP_INLINE_VISIBILITY _LIBCPP_CONSTEXPR_SINCE_CXX20
     static void construct(allocator_type&, _Tp* __p, _Args&&... __args) {
-#if _LIBCPP_STD_VER >= 20
```
Comment (EricWF) [inline libcxx/include/__memory/allocator_traits.h:303]: Same question as below.

Are we required to use `construct_at` here? If not, it seems to be more expensive, and causes arguably poorer diagnostic messages.

Comment (ldionne) [reply libcxx/include/__memory/allocator_traits.h:303]: This is for constexpr-friendliness IIRC.

**Comment by ldionne** [comment]
@EricWF This is essentially a small simplification of the code -- we have to use construct_at and destroy_at for constexpr-friendliness, and we also like to do it for consistency everywhere. So this really just trades `#if` for an unconditional call to something that does the right thing under the hood, so I'd go ahead with this change.

**Comment by EricWF** [comment]
What does "constexpr friendlyness" mean?
Is there magic in `construct_at`? 

Because it's just placement new.

If this adds constexpr friendlyness then doesn't in need more tests?

**Pull request merged**